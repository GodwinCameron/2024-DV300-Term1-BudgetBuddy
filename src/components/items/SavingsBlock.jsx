import React, { useEffect } from "react";

function SavingsBlock(props) {
  const { index, savings, savedPercentage, savedArray, setSavedArray } = props;
  console.log(props);

  useEffect(() => {

    setSavedArray &&
    setSavedArray((prevSavedArray) => {
      const updatedArray = [...prevSavedArray];
      const existingIndex = updatedArray.findIndex(
        (item) => item && item.index === index
      );

      if (existingIndex !== -1) {
        updatedArray[existingIndex] = {
          index: index,
          salary: savings.salary,
          saved: savings.salary * (savedPercentage / 100),
        };
      } else {
        updatedArray.push({
          index: index,
          salary: savings.salary,
          saved: savings.salary * (savedPercentage / 100),
        });
      }

      return updatedArray;
    });


  }, [savedPercentage]);

  return (
    <div data-testid="savings-block" className="card card-col p-2 row-savings">
      {savings && (
        <>
          <div className="card-row">
            <span>
              <div className="avatar">{savings.icon}</div>
              <p>{savings.name}</p>
            </span>
          </div>

          <p className="amount amount-save">
            {new Intl.NumberFormat("en-za", {
              style: "currency",
              currency: "ZAR",
            }).format(savings.salary * (savedPercentage / 100))}
          </p>
        </>
      )}
    </div>
  );
}

export default SavingsBlock;
