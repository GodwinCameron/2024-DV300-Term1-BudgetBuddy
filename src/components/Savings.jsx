import React from "react";
import TaxBlock from "./items/TaxBlock";
import { dummyIncome, percentageOptions } from "../utils";
import { Form } from "react-bootstrap";
import SavingsBlock from "./items/SavingsBlock";

function Savings(props) {
  const {
    savedPercentage,
    setSavedPercentage,
    income,
    savedArray,
    setSavedArray,
  } = props;

  const handleChange = (e) => {
    setSavedPercentage(e.target.value);
  };

  return (
    <div>
      <div className="title-row">
        <h3>Savings Calculation</h3>
        <span>
          <p>% you want to save</p>
          <Form.Select
            onChange={handleChange}
            id="percentage"
            name="percentage"
            defaultValue="-"
            autoComplete="off"
          >
            <option disabled>-</option>
            {percentageOptions.map((amount, index) => (
              <option key={index} value={amount}>
                {amount}%
              </option>
            ))}
          </Form.Select>
        </span>
      </div>

      {/* List */}
      <div className="scroll-row hide-scroll">
        {income &&
          income.map((item, index) => (
            <SavingsBlock
              savedPercentage={savedPercentage}
              key={index}
              index={index}
              savings={item}
              savedArray={savedArray}
              setSavedArray={setSavedArray}
            />
          ))}
      </div>
    </div>
  );
}

export default Savings;
