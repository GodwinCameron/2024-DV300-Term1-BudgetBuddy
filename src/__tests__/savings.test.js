import { render, screen, waitFor } from "@testing-library/react";
import Savings from "../components/Savings";
import SavingsBlock from "../components/items/SavingsBlock";

describe("Tests for savings", () => {
  test("test default state of savings compoenet without array", () => {
    render(<Savings />);

    var title = screen.getByText("Savings Calculation");

    expect(title).toBeInTheDocument();
    console.log(
      "this helps by both testing if the component can redner in the test environment",
      "for further tests and tests how the component handles not recieving the starting array"
    );
  });
});

describe("Test for savingsBlock component", () => {
  test("test the default state of the savingsBlock", () => {
    render(<SavingsBlock />);

    var savingsBlockContainer = screen.getByTestId("savings-block");

    expect(savingsBlockContainer).toBeInTheDocument();
  });
  test("test the savingsBlock component with basic savedArray", async () => {
    var index = 0;
    var savedPercentage = 10;
    var savings = { icon: "👱‍♂️", name: "John", salary: "12500" };
    var savedArray = [{ index: 0, salary: "12500", saved: 1250 }];
    const mockSetSavedArray = jest.fn();

    render(
      <SavingsBlock
        savedArray={savedArray}
        index={index}
        savedPercentage={savedPercentage}
        savings={savings}
        // setSavedArray={mockSetSavedArray}
      />
    );
    var savingsBlockContainer = screen.getByTestId("savings-block");
    var savingsBlockName = screen.getByText("John");
    var savingsBlockSaved = screen.getByText("R 1,250.00");

    // await waitFor(() => {
    //   expect(mockSetSavedArray).toHaveBeenCalledWith([
    //     {
    //       index: index,
    //       salary: savings.salary,
    //       saved: savings.salary * (savedPercentage / 100),
    //     },
    //   ]);

      expect(savingsBlockContainer).toBeInTheDocument();
      expect(savingsBlockName).toBeInTheDocument();
      expect(savingsBlockSaved).toBeInTheDocument();
    // });
  });
});

describe("snapshots", () => {
  test("snapshot of savings", () => {
    const { asFragment } = render(<Savings />);
    expect(asFragment()).toMatchSnapshot();
  });
});
