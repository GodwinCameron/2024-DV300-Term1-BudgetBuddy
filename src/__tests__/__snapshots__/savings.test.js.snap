// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots snapshot of savings 1`] = `
<DocumentFragment>
  <div>
    <div
      class="title-row"
    >
      <h3>
        Savings Calculation
      </h3>
      <span>
        <p>
          % you want to save
        </p>
        <select
          autocomplete="off"
          class="form-select"
          id="percentage"
          name="percentage"
        >
          <option
            disabled=""
            selected=""
          >
            -
          </option>
          <option
            value="5"
          >
            5%
          </option>
          <option
            value="10"
          >
            10%
          </option>
          <option
            value="15"
          >
            15%
          </option>
          <option
            value="20"
          >
            20%
          </option>
        </select>
      </span>
    </div>
    <div
      class="scroll-row hide-scroll"
    />
  </div>
</DocumentFragment>
`;
